### DISCUSSION TITLE
Run async operations on separate threads

### DISCUSSION THREAD

**Discussion by gonzacastro** [fastapi/fastapi#6305] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# some_router.py
@router.get("/")
async def get_some_things():
   return await some_long_taking_method()
```

### Description

Hey, I wanted to ask a question that has been buzzing me all over these days. It's about FastAPI and it behavior with threads and async/not async functions.
As you can see in the sample code, I have a router with an async endpoint that runs an operation that takes for example 15 seconds. After doing some research I found this: "Thus, def (sync) routes run in a separate thread from a threadpool, or, in other words, the server processes the requests concurrently, whereas async def routes run on the main (single) thread, i.e., the server processes the requests sequentially". So, to summarize, what I achieve currently is an endpoint that that takes some time to process and if 2 users hit it at the same time, one has to wait until the other finishes up... and that only speaking with only 2 users. Is there any way to wrap these endpoints to run on another thread? I tried using loads of things that some people suggested but none of them worked, so I anyone can give me a hand with it I would really appreciate it! Thanks in advance.

### Operating System

Windows

### FastAPI Version

0.79

### Python Version

3.9.13

**Comment by anthonycepeda** [comment]
Hi @gonzacastro,

Have you try adding [Uvicorn workers](https://fastapi.tiangolo.com/deployment/server-workers/#gunicorn-with-uvicorn-workers)?

For example let's say you have an entry point to your API like [this]( https://github.com/anthonycepeda/fastapi-sqlmodel/blob/main/asgi.py) but with an extra parameter: `workers=2` as  follows:

```python
uvicorn.run("asgi:api", host="0.0.0.0", port=8080, reload=True, workers=2)
```

**Comment by odiseo0** [comment]
@anthonycepeda `reload` and `workers` are parameters mutually exclusive, you can't really use `reload` & `workers` together.

But Uvicorn workers are a way to do that, anyway, If you're going to have long tasks I don't recommend waiting the response from the task

**Comment by gonzacastro** [comment]
@anthonycepeda Yeah I made an approach with that but there's some problem with uvicorn when setting up workers on Windows, it works when it's deployed on the cloud but not in my dev environment, so I'd like a better alternative.
@odiseo0 I know I'm not supposed to wait for too long responses but it is required because in that endpoint there are a few API calls to an external service that ends up being like 15 seconds depending some parameters. So I wanted to split up the work. The approach with the workers is okay but another thing I don't like is to set up the number of workers, I assume that the more workers I assign, the heavier is the job for the CPU, and I wouldn't like to set a few and take the risk to have a final user to be waiting for too long.

**Comment by iudeen** [comment]
Making your long running function to synchronous function may help.

**Comment by gonzacastro** [comment]
@iudeen I can't make it sync because it has an async/await within, so I have to mark all the way up to the router with async

**Comment by odiseo0** [comment]
How about making that ` await some_long_taking_method()` sync?

```py

# Code from https://github.com/tiangolo/asyncer/blob/main/asyncer/_main.py#L230

import functools
from typing import Any, Callable, TypeVar, Coroutine

import anyio
from anyio._core._eventloop import threadlocals
from typing_extensions import ParamSpec


T = TypeVar("T")
T_Retval = TypeVar("T_Retval")
T_ParamSpec = ParamSpec("T_ParamSpec")


def syncify(
    func: Callable[T_ParamSpec, Coroutine[Any, Any, T_Retval]],
    raise_sync_error: bool = True,
) -> Callable[T_ParamSpec, T_Retval]:
    """
    Take an async function and return a sync function that takes the same parameters as the original.
    """
    @functools.wraps(func)
    def wrapper(*args: T_ParamSpec.args, **kwargs: T_ParamSpec.kwargs) -> T_Retval:
        current_async_module = getattr(threadlocals, "current_async_module", None)
        partial_f = functools.partial(func, *args, **kwargs) # bind parameters here.

        if current_async_module is None and raise_sync_error is False:
            return anyio.run(partial_f)

        return anyio.from_thread.run(partial_f)
        
   sync_long_running_task = syncify(some_long_taking_method)   
   
   sync_long_running_task()
  ```
  
  Maybe it could help, but have you tried with `BackgroundTasks` too?

**Comment by iudeen** [comment]
Ideally async functions shouldn't have any blocking logic in it. 

I think making your async function sync is better approach - you may try @odiseo0 approach or change your function 😅

**Comment by JarroVGIT** [comment]
> I can't make it sync because it has an async/await within, so I have to mark all the way up to the router with async

But not in the right place then? If you have blocking code in your async function, then you will block the ~main thread~ event loop. That is not a FastAPI (or Python) 'issue' but an software design/implementation error. If your code awaits the longrunning calls to external API's, then FastAPI is able to handle other requests in the mean time, concurrently (but not parallel).

**Comment by gonzacastro** [comment]
@odiseo0 I liked that syncify method, I will try it right up, thanks! And regarding Background tasks, no, I haven't tried that because I require what that long lasting method returns, actually, what it does is collect data consuming an external API (which is awfully optimized so that's what takes too long) and returns is to the user. Therefore, as too many requests are made to that API, to reduce the times, I parallelized them and that's what it makes it an async operation.
@iudeen unfortunatelly it is required to do that, and I was told to fix that performance issue when several clients hit the async endpoint.
@JarroVGIT actually I don't block the main thread, I explained what the endpoint is supposed to do above, do you still think it is an implementation issue? Because I have worked with other APIs frameworks in for example Go and each request would always be handled on a exclusive thread. And now, in my currrent project we're using Python, so I have to figure out how to get that.

**Comment by JarroVGIT** [comment]
@gonzacastro My apologies, I hadn't had my morning coffee yet. I meant that when you have synchronous blocking code in an async call stack, it will result in blocking the **event loop** (not thread, like i said earlier). That is, in my opinion, a design flaw of the software. You can either:
- Use synchronous code (and sync path operation function), so FastAPI will run it in a seperate thread, and Python will yield control back when possible (e.g. when waiting for IO), or;
- Use asynchronous code (and async path operation function), so FastAPI will run it in the event loop, and your code will yield control back when possible.

If your code (and I mean the 15-second-runtime-code) is not IO bound but CPU bound (e.g. you are performing some crazy calculation), then again I would recommend to reconsider your design and make the entire request-response loop more asynchronisous. With that, I mean going from `request -> calculation -> return response` to `request -> put work in queue and return response` so your client can check if the result is available later on. However, if it is IO bound (the 15 second work), you should either make the IO bound work async, or use sync all the way. Mixing them both in a IO heavy situation (at least in FastAPI) is a bad idea and should be avoided.

**Comment by mcleantom** [comment]
Asyncronous code is different to multi-threaded code:

> Classical computer science emphasizes efficient algorithms that complete computations as quickly as possible. But many networked programs spend their time not computing, but holding open many connections that are slow, or have infrequent events. These programs present a very different challenge: to wait for a huge number of network events efficiently. A contemporary approach to this problem is asynchronous I/O, or "async".
> 
> This chapter presents a simple web crawler. The crawler is an archetypal async application because it waits for many responses, but does little computation. The more pages it can fetch at once, the sooner it completes. If it devotes a thread to each in-flight request, then as the number of concurrent requests rises it will run out of memory or other thread-related resource before it runs out of sockets. It avoids the need for threads by using asynchronous I/O

[A Web Crawler With asyncio Coroutines
A. Jesse Jiryu Davis and Guido van Rossum](http://aosabook.org/en/500L/a-web-crawler-with-asyncio-coroutines.html)

So the benefit of asyncronous versus threaded code depends on what `some_long_taking_method()` is doing.

If `some_long_taking_method()` is doing I/O bound tasks (reading from files, doing http requests), then async is (probably) better than running in a thread pool.

If `some_long_taking_method()` is heavy computation, then you can run it in a separate threadpool (which is automatically done by FastAPI if you do `def` rather than `async def`), however you need to make sure that the code can be multithreaded (i.e. the C/C++ implementation releases the GIL)

Alternatively, [you can run the computation in a separate process ](https://stackoverflow.com/questions/63169865/how-to-do-multiprocessing-in-fastapi)

**Comment by gonzacastro** [comment]
@mcleantom what I do in that function is a lot of http requests, so what i did was parallelize all of them and gather them up when finished. What do you recommend in that case?

**Comment by mcleantom** [comment]
> @mcleantom what I do in that function is a lot of http requests, so what i did was parallelize all of them and gather them up when finished. What do you recommend in that case?

use asyncio.gather with an asynchronous requests library (httpx)

**Comment by iudeen** [comment]
> @mcleantom what I do in that function is a lot of http requests, so what i did was parallelize all of them and gather them up when finished. What do you recommend in that case?

Using Async is not parallelizing, it's concurrency. If you need parallelizing, check threadpool or even process pool.

**Comment by gonzacastro** [comment]
@mcleantom but in that case i will have to put the await clause as well because the asyncio gather requires it.
@iudeen I use the method to_thread of asyncio, isn't that what you mean? Because that's what I do inside the method

**Comment by JarroVGIT** [comment]
To be frank, I would recommend to read up on the topics “threads”, “async” and “GIL” to understand the differences and how they compare to each other and their relations. You can never have parallelism in one python process, only concurrency. If you have none-asynchronous blocking IO, use threads (and sync endpoints). If you have asynchronous blocking IO, use async (and async endpoints). 

Calling multiple http endpoints is a perfect example of something that could easily be made async (with httpx for example). That way, while one request is waiting for an external API to return, FastAPI will handle other request in the mean time.

**Comment by JarroVGIT** [comment, marked as answer]
To start off, this issue makes me proud being part of this community! Great help from all over the place, that is really cool to see!

I had a brief conversation on LinkedIn with @gonzacastro , and I now understand what he is trying to achieve. I created an example that demonstrates that the FastAPI can still be requested, while other requests are still waiting for the 3rd party API to respond. 

```python
import asyncio
import random
import time

from fastapi import FastAPI, Request

app = FastAPI()

# keep a count of requests, makes reading the log easier.
app.state.request_no = 0


async def get_response_from_external_api(url: str, request_no) -> None:
    # here you would put something like httpx to make async calls to your endpoint.
    # to simulate this takes a while, we will sleep here. And the return value is the
    # url, the time it should take (sleeptime) and the time it did take.
    sleep_time = random.randint(1, 5)
    print(f"({request_no}) Calling {url}, should take {sleep_time} seconds...")
    start = time.time()
    await asyncio.sleep(sleep_time)
    end = time.time()
    print(f"({request_no}) Done calling {url}, took {end-start} seconds...")
    return (url, sleep_time, end - start)


@app.get("/get_response")
async def get_response(r: Request):
    # Using app.state because it will change with every incoming request
    r.app.state.request_no += 1
    request_no = r.app.state.request_no

    # Construct the URLs
    urls = [f"url.com/api/{i}" for i in range(1, 5)]
    print(f"({request_no}) Request has come in.")

    start = time.time()
    result = await asyncio.gather(
        *[get_response_from_external_api(url, request_no) for url in urls]
    )

    end = time.time()
    print(f"({request_no}) Request has been processed.")
    return result


@app.get("/")
async def root():
    return {"hello": "world"}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```

I added some verbosity for clarity purposes. If I call the `get_response` endpoint 3 times rapidly, the logs look like this:
```log
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
(1) Request has come in.
(1) Calling url.com/api/1, should take 3 seconds...
(1) Calling url.com/api/2, should take 1 seconds...
(1) Calling url.com/api/3, should take 3 seconds...
(1) Calling url.com/api/4, should take 5 seconds...
(2) Request has come in.
(2) Calling url.com/api/1, should take 2 seconds...
(2) Calling url.com/api/2, should take 2 seconds...
(2) Calling url.com/api/3, should take 2 seconds...
(2) Calling url.com/api/4, should take 2 seconds...
(3) Request has come in.
(3) Calling url.com/api/1, should take 5 seconds...
(3) Calling url.com/api/2, should take 4 seconds...
(3) Calling url.com/api/3, should take 1 seconds...
(3) Calling url.com/api/4, should take 3 seconds...
(1) Done calling url.com/api/2, took 1.0014400482177734 seconds...
(3) Done calling url.com/api/3, took 1.0007548332214355 seconds...
(2) Done calling url.com/api/1, took 2.0016140937805176 seconds...
(2) Done calling url.com/api/2, took 2.0016257762908936 seconds...
(2) Done calling url.com/api/3, took 2.0016050338745117 seconds...
(2) Done calling url.com/api/4, took 2.0015769004821777 seconds...
(2) Request has been processed.
INFO:     127.0.0.1:56387 - "GET /get_response HTTP/1.1" 200 OK
(1) Done calling url.com/api/1, took 3.0016939640045166 seconds...
(1) Done calling url.com/api/3, took 3.0017738342285156 seconds...
(3) Done calling url.com/api/4, took 3.001483201980591 seconds...
(3) Done calling url.com/api/2, took 4.001469850540161 seconds...
(1) Done calling url.com/api/4, took 5.001488208770752 seconds...
(1) Request has been processed.
INFO:     127.0.0.1:56386 - "GET /get_response HTTP/1.1" 200 OK
(3) Done calling url.com/api/1, took 5.001416921615601 seconds...
(3) Request has been processed.
INFO:     127.0.0.1:56388 - "GET /get_response HTTP/1.1" 200 OK
```

Note how request 1 took longer than request 2, and although request 2 came in later than request 1, it returned it response earlier. 

You just have to change the `asyncio.sleep()` into a asynchronous api call (use `httpx`) and your good to go. Hope this answers your question!

**Comment by gonzacastro** [comment]
@JarroVGIT thank you very much, all things clear!
